pkgs/fc/check-ceph/nautilus/fc/check_ceph/check_snapshot_restore.py

```diff
@@ -0,0 +1,251 @@
+"""Check that the largest RBD image snapshot of each Ceph cluster root can be restored,
+without exceeding the near_full threshold of the cluster capacity.
+Please consider that this is not a guarantee for not getting `OSD_NEARFULL` warnings,
+as individual fill levels of OSDs below a cluster root is not perfectly balanced."""
+
+import contextlib
+import json
+import sys
+from collections import namedtuple
+from time import sleep
+
+import rados
+import rbd
+import toml  # replace this with tomllib when on python-3.11+
+
+KiB = 1024
+MiB = 1024 * KiB
+GiB = 1024 * MiB
+
+EXIT_OK = 0
+EXIT_WARN = 1
+EXIT_CRITICAL = 2
+EXIT_UNKNOWN = 3
+
+EXIT_PRIORITIES = [
+    EXIT_CRITICAL,
+    EXIT_WARN,
+    EXIT_UNKNOWN,
+    EXIT_OK,
+]
+
+Thresholds = namedtuple("thresholds", ["nearfull", "full"])
+
+
+def main():
+    # can cause the program to exit with non-working config
+    (thresholds, pool_roots) = parse_config(sys.argv)
+    with cluster_connection() as cluster:
+        fill_stats = get_cluster_fillstats(cluster, pool_roots)
+        largest_snaps = get_largest_snaps(cluster, pool_roots)
+
+    sys.exit(eval_fill_warnings(fill_stats, largest_snaps, thresholds))
+
+
+def parse_config(argv) -> (Thresholds, dict):
+    if not len(argv) == 2:
+        print(f"Usage: {argv[0]} <config.toml>")
+        sys.exit(EXIT_CRITICAL)
+
+    try:
+        with open(argv[1], "rt") as configtoml:
+            config = toml.load(configtoml)
+    except (FileNotFoundError, PermissionError):
+        print(f"Error: Unable to open file {argv[1]}.")
+        sys.exit(EXIT_CRITICAL)
+    except toml.TomlDecodeError:
+        print(f"Error: {argv[1]} is not a valid TOML file.")
+        sys.exit(EXIT_CRITICAL)
+
+    # repacking the required values serves as an implicit schema validation
+    try:
+        thresholds = Thresholds(
+            nearfull=float(config["thresholds"]["nearfull"]),
+            full=float(config["thresholds"]["full"]),
+        )
+        pool_roots = config["ceph_roots"]
+        for root, pools in pool_roots.items():
+            assert len(pools) > 0
```

Comment:
this causes a crash and will reduce visibility, however, maybe that's acceptable because if that happens we need to (timely) fix that anyway. let's leave as it is.